### PR TITLE
get_class/get_class_at implemented

### PR DESCRIPTION
**Pull request by taco-paco**
## Usage related changes

#61 + #59 

## Development related changes
models/CotractClass united with types/ContractClass

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [ ] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: rustfmt.toml [hunk 1]
```diff
@@ -4,17 +4,3 @@ use_field_init_shorthand = true
 use_small_heuristics = "Max"
 use_try_shorthand = true
 max_width = 100
-
-# Unstable features below
```
Comment (FabijanC) [inline rustfmt.toml:8]: Doesn't this unnecessarily increase the changed-line count of this PR? Can this be handled in a separate PR to make review easier?

File: crates/starknet-server/src/api/json_rpc/mod.rs [hunk 2]
```diff
@@ -128,6 +128,11 @@ impl<T: Serialize> ToRpcResponseResult for RpcResult<T> {
                     message: err.to_string().into(),
                     data: None,
                 },
+                ApiError::SerializationError(error) => RpcError {
```
Comment (FabijanC) [inline crates/starknet-server/src/api/json_rpc/mod.rs:131]: Shouldn't this kind of error be a signal of a bug in our code?
Anyway, this way we will be displaying a "Serialization error" message to users, without any explanation and with a potential debugging difficulty.

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 3]
```diff
@@ -232,26 +241,21 @@ impl JsonRpcHandler {
         block_id: BlockId,
         contract_address: ContractAddress,
     ) -> RpcResult<ClassHash> {
-        let starknet = self.api.starknet.read().await;
-        match starknet.get_class_hash_at(&block_id.into(), &contract_address) {
+        match self.api.starknet.write().await.get_class_hash_at(block_id.into(), contract_address) {
```
Comment (FabijanC) [inline crates/starknet-server/src/api/json_rpc/endpoints.rs:244]: Hear and in other places (`get_class`, `get_class_at`), why lock this with `write` privilege instead of read?

File: rustfmt.toml
Comment (marioiordanov) [inline rustfmt.toml:1]: why the unstable features are removed

**Review by marioiordanov — CHANGES_REQUESTED** [review]
Please merge main branch, because the ContractAddressHex -> ContractAddress is already merged in main. IMO this will reduce the number of files changed in the PR.

File: rustfmt.toml (hunk 1 above)
Comment (taco-paco) [reply rustfmt.toml]: Were tmp changes that I forgot to undo

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 3 above)
Comment (taco-paco) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: Old implementation tapped in CachedState which used &mut state for `get_class_hash_at`. Now reverted to read access

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 4]
```diff
@@ -196,10 +196,25 @@ impl JsonRpcHandler {
...
     pub(crate) async fn get_class(
         &self,
-        _block_id: BlockId,
-        _class_hash: ClassHash,
-    ) -> RpcResult<ContractClass> {
-        Err(error::ApiError::ClassHashNotFound)
+        block_id: BlockId,
+        class_hash: ClassHash,
+    ) -> RpcResult<CodegenContractClass> {
+        Ok(self.api.starknet.read().await.get_class(block_id.into(), class_hash)?.try_into()?)
```
Comment (marioiordanov) [inline crates/starknet-server/src/api/json_rpc/endpoints.rs:202]: Why there is no error handling in this method like block not found, class hash not found. This way will use the common StarknetDevnetError

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 5]
```diff
@@ -208,26 +223,21 @@ impl JsonRpcHandler {
...
             Err(Error::NoBlock) => Err(ApiError::BlockNotFound),
-            Err(Error::ContractNotFound | Error::NoStateAtBlock { block_number: _ }) => {
-                Err(ApiError::ContractNotFound)
+            Err(
+                Error::ContractNotFound
+                | Error::StateError(StateError::NoneContractState(_))
+                | Error::NoStateAtBlock { block_number: _ },
+            ) => Err(ApiError::ContractNotFound),
+            Err(unknown_error) => {
+                let unknown_error = unknown_error;
```
Comment (marioiordanov) [inline crates/starknet-server/src/api/json_rpc/endpoints.rs:235]: Why this line is left, maybe because of debugging
```suggestion
```

File: crates/starknet-server/src/api/json_rpc/error.rs [hunk 6]
```diff
@@ -38,6 +38,8 @@ pub enum ApiError {
     InvalidContractClass,
     #[error("Only latest/pending block is supported")]
     OnlyLatestBlock,
+    #[error("Serialization error")]
+    SerializationError(#[from] serde_json::Error),
```
Comment (marioiordanov) [inline crates/starknet-server/src/api/json_rpc/error.rs:42]: The error is not used anywhere.

File: crates/starknet-server/tests/common/mod.rs [hunk 7]
```diff
@@ -145,8 +145,8 @@ pub mod util {
 
             let process = Command::new("cargo")
                 .arg("run")
-                .arg("--release")
                 .arg("--")
```
Comment (marioiordanov) [inline crates/starknet-server/tests/common/mod.rs:148]: using this order of arguments an error arises: "error: unexpected argument '--release' found"

File: crates/types/src/abi_entry.rs
Comment (marioiordanov) [inline crates/types/src/abi_entry.rs:1]: why not move abi_entry.rs into contract_class/deprecated

File: crates/starknet/src/error.rs
Comment (marioiordanov) [inline crates/starknet/src/error.rs:1]: those errors are not used anywhere

File: crates/starknet-server/tests/common/mod.rs (hunk 7 above)
Comment (taco-paco) [reply crates/starknet-server/tests/common/mod.rs:148]: My bad screwed up with [revert](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/126/commits/a93bd4078161ade5683e5cf1581cedced25972e7)

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 4 above)
Comment (taco-paco) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:202]: Seems weird that we have to do all of that. Why don't we use [StarknetError](https://github.com/xJonathanLEI/starknet-rs/blob/e9bdd7efeae9c76ef62a130ccef3f4355a864797/starknet-core/src/types/codegen.rs#L1026) from `starknet-core`? It has all error & error codes defined & display implemented. Also our errors would be compliant with starknet-rs

Comment (taco-paco) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:202]: Done

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 8]
```diff
@@ -196,10 +196,25 @@ impl JsonRpcHandler {
...
+        Ok(self.api.starknet.read().await.get_class(block_id.into(), class_hash)?.try_into()?)
+    }
+
+    /// starknet_getClassAt
+    pub(crate) async fn get_class_at(
+        &self,
+        block_id: BlockId,
+        contract_address: ContractAddress,
+    ) -> RpcResult<CodegenContractClass> {
+        Ok(self
```
[the remark this answers is not in the document]
Comment (taco-paco) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:211]: Done

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 5 above)
Comment (taco-paco) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:235]: Yeap. Removed

File: crates/types/src/contract_class/deprecated/rpc_contract_class.rs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        Ok(starknet_in_rust_contract_class)
+    }
+}
+
+impl TryInto<CompressedLegacyContractClass> for DeprecatedContractClass {
+    type Error = Error;
+    fn try_into(self) -> Result<CompressedLegacyContractClass, Self::Error> {
+        // TODO: improve
+        let value: Cairo0Json = self.try_into()?;
+        Ok(serde_json::from_value(value.inner).map_err(JsonError::SerdeJsonError)?)
```
[the remark this answers is not in the document]
Comment (taco-paco) [reply crates/types/src/contract_class/deprecated/rpc_contract_class.rs:99]: `Cairo0Json` is a wrapper around `serde_json::Value`. `value.inner` has type `serde_json::Value`

File: crates/starknet-server/src/api/json_rpc/error.rs (hunk 6 above)
Comment (taco-paco) [reply crates/starknet-server/src/api/json_rpc/error.rs:42]: Yeap, not anymore, thanks

File: crates/types/src/abi_entry.rs
Comment (taco-paco) [reply crates/types/src/abi_entry.rs:1]: Good catch. Done

File: crates/types/src/contract_class/deprecated/rpc_contract_class.rs [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        deserialize_with = "deserialize_to_serde_json_value_with_keys_ordered_in_alphabetical_order",
+        serialize_with = "serialize_program_to_base64"
+    )]
+    pub program: Value,
+    pub abi: Vec<ContractClassAbiEntryWithType>,
+    /// The selector of each entry point is a unique identifier in the program.
+    pub entry_points_by_type: LegacyEntryPointsByType,
+}
+
+impl DeprecatedContractClass {
```
[the remark this answers is not in the document]
Comment (taco-paco) [reply crates/types/src/contract_class/deprecated/rpc_contract_class.rs:39]: Had it in local tests, added. Also would be nice to keep anyway in case we add artefacts with zipped program

File: crates/starknet/src/error.rs [hunk 11]
```diff
@@ -11,6 +11,12 @@ pub enum Error {
     TransactionError(#[from] starknet_in_rust::transaction::error::TransactionError),
     #[error("Types error")]
     TypesError(#[from] starknet_types::error::Error),
+    #[error("Specifying block by hash is currently not enabled")]
+    BlockIdHashUnimplementedError,
+    #[error("Specifying block by number is currently not enabled")]
+    BlockIdNumberUnimplementedError,
+    #[error("Contract hash not found")]
+    ContractHashNotFound,
```
[the remark this answers is not in the document]
Comment (taco-paco) [reply crates/starknet/src/error.rs:19]: Removed

File: crates/types/src/serde_helpers.rs
[the remark this answers is not in the document]
Comment (taco-paco) [reply crates/types/src/serde_helpers.rs:1]: Moved back

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 4 above)
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: sure, created an issue for it https://github.com/0xSpaceShard/starknet-devnet-rs/issues/130

File: crates/types/src/contract_class/deprecated/rpc_contract_class.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/types/src/contract_class/deprecated/rpc_contract_class.rs]: yes, but `CompressedLegacyContractClass` contains byte array for `program` and in the json format the field is an object

**Review by marioiordanov — APPROVED** [review]
